File: fn/src/main/java/org/sqlbuilder/fn/joins/FEGroupPattern_AnnoSet.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.fn.RequiresIdFrom;

import java.util.HashSet;
import java.util.Set;

import edu.berkeley.icsi.framenet.AnnoSetType;

/*
grouppatterns_annosets.table=fnpatterns_annosets
grouppatterns_annosets.create=CREATE TABLE IF NOT EXISTS %Fn_patterns_annosets.table% ( patternid INTEGER NOT NULL,annosetid INTEGER NOT NULL,PRIMARY KEY (patternid,annosetid) );
grouppatterns_annosets.fk1=ALTER TABLE %Fn_patterns_annosets.table% ADD CONSTRAINT fk_%Fn_patterns_annosets.table%_patternid FOREIGN KEY (patternid) REFERENCES %Fn_patterns.table% (patternid);
grouppatterns_annosets.fk2=ALTER TABLE %Fn_patterns_annosets.table% ADD CONSTRAINT fk_%Fn_patterns_annosets.table%_annosetid FOREIGN KEY (annosetid) REFERENCES %Fn_annosets.table% (annosetid);
grouppatterns_annosets.no-fk1=ALTER TABLE %Fn_patterns_annosets.table% DROP CONSTRAINT fk_%Fn_patterns_annosets.table%_annosetid CASCADE;
grouppatterns_annosets.no-fk2=ALTER TABLE %Fn_patterns_annosets.table% DROP CONSTRAINT fk_%Fn_patterns_annosets.table%_patternid CASCADE;
grouppatterns_annosets.insert=INSERT INTO %Fn_patterns_annosets.table% (patternid,annosetid) VALUES(?,?);
 */
public class FEGroupPattern_AnnoSet extends Pair<FEGroupPattern, Integer> implements Insertable<FEGroupPattern_AnnoSet>
{
	public static final Set<FEGroupPattern_AnnoSet> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static FEGroupPattern_AnnoSet make(final FEGroupPattern pattern, final AnnoSetType annoset)
	{
		var pa = new FEGroupPattern_AnnoSet(pattern, annoset.getID());
		SET.add(pa);
		return pa;
	}

	private FEGroupPattern_AnnoSet(final FEGroupPattern pattern, final int annosetid)
	{
		super(pattern, annosetid);
	}

	// I N S E R T A B L E

	@RequiresIdFrom(type = FEGroupPattern.class)
	@Override
	public String dataRow()
	{
		return String.format("%s,%d", first.getSqlId(), second);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[PAT-AS pattern=%s annosetid=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/FEGroupPattern_AnnoSet.kt
package org.sqlbuilder.fn.joins

import edu.berkeley.icsi.framenet.AnnoSetType
import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.Insertable
import java.util.*

data class FEGroupPattern_AnnoSet(
    val pattern: FEGroupPattern,
    val annosetid: Int,
) : Insertable {

    // I N S E R T A B L E

    @RequiresIdFrom(type = FEGroupPattern::class)
    override fun dataRow(): String {
        return "${pattern.getSqlId()},$annosetid"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FEGroupPattern_AnnoSet
        return pattern == that.pattern && annosetid == that.annosetid
    }

    override fun hashCode(): Int {
        return Objects.hash(pattern, annosetid)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[PAT-AS pattern=$pattern annosetid=$annosetid]"
    }

    companion object {

        val SET = HashSet<FEGroupPattern_AnnoSet>()

        fun make(pattern: FEGroupPattern, annoset: AnnoSetType): FEGroupPattern_AnnoSet {
            val pa = FEGroupPattern_AnnoSet(pattern, annoset.getID())
            SET.add(pa)
            return pa
        }
    }
}
